Load replay files asynchronously for the leaderboards

bltech opened the pull request:
Closes #930 

The logic that looks for replay files now dispatches asynchronous file read requests to the filesystem. As soon as the callback is called, the replay is added to the list and the leaderboard gets updated accordingly.


~~NOTE: I'm not really sure if what I'm doing with the raw buffer the filesystem provides causes a memory leak. If I interpret it right, the CUtlBuffer should take ownership of the buffer to free it at some point but if I do that, it blocks and the game freezes for a short while for some reason. I looked at the RAM and even with 153 replays I couldn't find any increase when loading everything repeatedly.~~ Not an issue any more

### Checklist
- [x] **I have thoroughly tested all of the code I have modified/added/removed to ensure something else did not break**
- [x] If there is a localization token change, I have updated the `momentum_english_ref.res` file with the changes, ran `tokenizer.py` to generate an up-to-date localization file, and have committed both the `.res` file changes and the new localization `.txt` file
- [x] If I introduced new h/cpp files, I have added them to the appropriate project's VPC file (`server_momentum.vpc` / `client_momentum.vpc` / etc)
- [x] If I have added or modified any visual assets (models, materials, panels, effects, etc), I have taken screenshots / videos of them and attached them to this PR directly (screenshots uploaded through github, videos uploaded to youtube and linked)
- [x] If I have modified any console command, console variable, or momentum entity, I have opened an issue (or a PR) for it in the [Momentum Mod documentation repository](https://github.com/momentum-mod/docs)
- [x] My commits are relatively small and scoped to the best of my ability
- [x] My branch has a clear history of changes that can be easy to follow when being reviewed commit-by-commit
- [x] My branch is functionally complete; the only changes to be done will be those potentially requested in code review

<!-- If any of these items are giving you doubts, please ask about it in the Discord! -->

Gocnak commented on mp/src/game/client/momentum/ui/leaderboards/LeaderboardsTimes.cpp:
```diff
@@ -365,27 +383,37 @@ void CLeaderboardsTimes::LoadLocalTimes(KeyValues* kv)
+    auto replay = g_ReplayFactory.LoadReplayFromBuffer(reader, false);
+    if(replay)
+    {
+        m_vLocalTimes.Insert(replay);
```
I worry about race conditions surrounding this.

What happens if the user hits tab to load these and leaves the map?

Is there any time where this vector would be cleared while being filled asynchronously? Since there's no mutex, is this thread safe?

bltech replied:
Yeah, I guess so. I think I will add another vector that stores all pending async request handles and aborts them in the panel's destructor. That should solve the first one, right?

For the second one a queue that gets filled by the callback and emptied when filling the scoreboard should do the trick?

bltech replied:
And while emptying the queue I should probably also suspend all async io

Gocnak replied:
Yeah I figured AsyncRead returns some sort of identifier to do the cancelling, but the panel isn't destroyed, the level is shutdown. There should be a hook for it, if there's not, we may want to pull async local map replay loading into its own system so that the map selector can also hook into it for its parsing it could do.

bltech replied:
Pulling it out seems like a good option, this would belong in the Replay Factory then, right?

Gocnak replied:
Yeah, as that's shared code so it'd be able to be used in the server as well if need be, but can for now exist under a `#ifdef CLIENT_DLL` block inside of it since only panels are showing loaded times.

Gocnak commented on mp/src/game/client/momentum/ui/leaderboards/LeaderboardsTimes.cpp:
```diff
@@ -365,27 +383,37 @@ void CLeaderboardsTimes::LoadLocalTimes(KeyValues* kv)
+    if(replay)
+    {
+        m_vLocalTimes.Insert(replay);
+        PostMessage(GetVParent(), new KeyValues("LocalTimeProcessed"));
```
Should it be to the parent or this panel?

bltech replied:
It is supposed to be the parent, so it can call Update and deal with the update timer on itself but it could also be the panel itself and only call FillLeaderboards, IMO the first one is cleaner

Gocnak replied:
Ah gotcha, all good then 👍

Gocnak commented on mp/src/game/client/momentum/ui/leaderboards/LeaderboardsTimes.cpp:
```diff
@@ -529,21 +557,20 @@ void CLeaderboardsTimes::OnlineTimesVectorToLeaderboards(TimeType_t type)
+    if (!m_bTimesNeedUpdate[TIMES_LOCAL])
+    {
+        float lastUp = gpGlobals->curtime - m_flTimesLastUpdate[TIMES_LOCAL];
+        m_bTimesNeedUpdate[TIMES_LOCAL] = fullUpdate && lastUp >= UPDATE_INTERVAL;
```
Shouldn't this be `||`?

Gocnak replied:
Wait nevermind, that would cause this to reread all the runs every `UPDATE_INTERVAL` seconds

Gocnak commented on mp/src/game/client/momentum/ui/leaderboards/LeaderboardsTimes.cpp:
```diff
@@ -529,21 +557,20 @@ void CLeaderboardsTimes::OnlineTimesVectorToLeaderboards(TimeType_t type)
 
     if (!g_pGameModeSystem->GameModeIs(GAMEMODE_UNKNOWN))
     {
-        for (int i = 1; i < TIMES_COUNT; i++) // Skip over local
```
This comment was okay because it explained why the loop starts at 1

bltech replied:
Oops, I originally changed how that worked but I forgot to re-add the comment!

Gocnak commented on mp/src/game/client/momentum/ui/leaderboards/LeaderboardsTimes.cpp:
```diff
@@ -958,6 +983,7 @@ void CLeaderboardsTimes::ApplySchemeSettings(IScheme* pScheme)
     m_cFirstPlace = pScheme->GetColor("FirstPlace", Color(240, 210, 147, 50));
     m_cSecondPlace = pScheme->GetColor("SecondPlace", Color(175, 175, 175, 50));
     m_cThirdPlace = pScheme->GetColor("ThirdPlace", Color(205, 127, 50, 50));
+    m_cItemDefault = pScheme->GetColor("Panel.BgColor", Color(54, 54, 54, 255));
```
The only worry I have with this is this BG color can go out of sync with this panel's bg color if it gets overridden. It might be best to define this also in the res file for this panel at least, and to default it to `GetBgColor()` in apply scheme settings.

bltech replied:
This is no longer an issue, the colors will no longer be re-set constantly

Gocnak commented on mp/src/game/client/momentum/ui/leaderboards/LeaderboardsTimes.h:
```diff
@@ -3,6 +3,7 @@
 #include "vgui_controls/EditablePanel.h"
 #include "steam/isteamhttp.h"
 #include "mom_shareddefs.h"
+#include "filesystem.h"
```
Is it possible to forward declare the things needed from this? Filesystem could be a pretty big include to have in this header

bltech replied:
possibly yes, I'm not sure though what the best way to forward-declare the enum is. I read you can do that with c++11 but GCC doesn't seem to like it

Gocnak replied:
Then no worries, it can be figured out in the future when the leaderboards are expanded upon in 0.11.0.

Gocnak commented on mp/src/game/shared/momentum/run/mom_replay_factory.cpp:
```diff
@@ -70,6 +71,28 @@ CMomReplayBase* CMomReplayFactory::LoadReplayFile(const char* pFileName, bool bF
+    return LoadReplayFromBuffer(reader, bFullLoad);
+}
+
+CMomReplayBase* CMomReplayFactory::LoadReplayFromBuffer(CUtlBuffer& reader, bool bFullLoad)
```
Can the buffer be passed as a const reference?

bltech replied:
not without changing CMomReplayV1, at least

braem commented on mp/src/game/client/momentum/ui/leaderboards/LeaderboardsTimes.cpp:
```diff
@@ -365,27 +383,37 @@ void CLeaderboardsTimes::LoadLocalTimes(KeyValues* kv)
+            request.pContext = this;
+            request.pfnCallback = LocalTimesCallbackForwarder;
+            request.pszPathID = "MOD";
+            filesystem->AsyncRead(request);
```
Minor, but looks like this is expecting 2 arguments:
![image](https://user-images.githubusercontent.com/9014762/96934577-4cf94380-1477-11eb-877b-0f0175067e57.png)

bltech replied:
Yeah, if I am going to store the handles like I said above, this will be fixed

braem commented on mp/src/game/client/momentum/ui/leaderboards/LeaderboardsTimes.cpp:
```diff
@@ -325,9 +335,17 @@ void CLeaderboardsTimes::SetPlaceColors(SectionedListPanel* panel, TimeType_t ty
             if (itemCount > 2)
+            {
                 panel->SetItemBgColor(panel->GetItemIDFromRow(2), m_cThirdPlace);
+                if(itemCount > 3)
```
Also minor, but we usually space this out. Eg. (`if ()`, `for ()`).

bltech replied:
I should probably point CLion towards the clangformat file but I'll also do another style pass including OVERRIDE etc.

bltech commented:
I only did the leaderboards for now. If you want, I can do the rest after this has reached a consensus (but maybe in another PR to not bloat this one too much?).

Gocnak commented on mp/src/game/client/momentum/ui/leaderboards/LeaderboardsTimes.cpp:
```diff
@@ -264,38 +265,17 @@ int CLeaderboardsTimes::TryAddAvatar(const uint64 &steamid, CUtlMap<uint64, int>
     if (m_pCurrentLeaderboards && m_pTop10Leaderboards && m_pAroundLeaderboards
         && m_pLocalLeaderboards && m_pFriendsLeaderboards)
     {
-        SetVisible(false);
```
You can remove the `SetVisible(true)` that went with this at the bottom of this func.

Gocnak commented on mp/src/game/client/momentum/ui/leaderboards/LeaderboardsTimes.cpp:
```diff
@@ -313,6 +293,29 @@ void CLeaderboardsTimes::FillLeaderboards(bool bFullUpdate)
+    SetVisible(false);
+    KeyValuesAD kvLocalTimes("local");
+    ConvertLocalTimes(kvLocalTimes);
+    if (kvLocalTimes && !kvLocalTimes->IsEmpty())
```
Since you now create the local times KV you can get rid of this null check and early return:
```suggestion
    if (kvLocalTimes->IsEmpty())
        return;
```

Gocnak commented on mp/src/game/client/momentum/ui/leaderboards/LeaderboardsTimes.cpp:
```diff
@@ -358,34 +361,33 @@ void CLeaderboardsTimes::LoadLocalTimes(KeyValues* kv)
-            m_vLocalTimes.RedoSort();
-            m_bTimesNeedUpdate[TIMES_LOCAL] = false;
-        }
+        auto callback = UtlMakeDelegate(this, &CLeaderboardsTimes::LocalTimesCallback);
+        g_ReplayFactory.AsyncLoadReplayFileMultiple(filenames, "MOD", callback, &m_mapReplayLocalRequest);
```
Callback variable isn't used anywhere so you can just do this:
```suggestion
        g_ReplayFactory.AsyncLoadReplayFileMultiple(filenames, "MOD", UtlMakeDelegate(this, &CLeaderboardsTimes::LocalTimesCallback), &m_mapReplayLocalRequest);
```

Gocnak commented on mp/src/game/client/momentum/ui/leaderboards/LeaderboardsTimes.cpp:
```diff
@@ -358,34 +361,33 @@ void CLeaderboardsTimes::LoadLocalTimes(KeyValues* kv)
         V_FixSlashes(path);
 
         FileFindHandle_t found;
+        CUtlStringList filenames;
```
I think this entire block of code of finding the filenames and adding it to a CUtlStringList should move into the AsyncLoadReplayFileMultiple method.

Mainly because there needs to be a key into this system so it knows to not step on top of itself. What I'm thinking instead is you pass the mapname in as the first parameter, and in an internal CUtlDict, tie that map name to the request handle. If the handle exists, don't create the new Async request. If it doesn't, do this file searching and then the request.

bltech replied:
If getting all replays for a single map is really the only use case we need to worry about then I'll add the map name as the first parameter and rename the method accordingly. 

If however this should be a bit more general, maybe the parameter could also be one string the caller constructs that then uses the wildcard-functionality of FindFirstEx in the factory.

Also, could you maybe elaborate a bit more about what you mean with stepping on top of itself? Do you mean the leaderboards generating requests multiple times? I'm not sure I get exactly what you are saying.

Gocnak replied:
I think I see what you mean here. The use case is going to be loading a specific file (map finished panel, run submitter, leaderboards on the `run_save` event) and loading all replays for a map (leaderboards Local tab).

In this case, being able to have a `LoadReplaysForMap(mapName, . . .)` that funnels into a `LoadReplaysByFilenameAsync(filenames, . . .)` would work because then the single-file use case can just call that with the one replay path it knows it has.

Gocnak replied:
As for stepping on toes, yeah it would just be a matter of preventing the exact same request being done, though I suppose this is a very explicit API so maybe it would be better to debug if it gets called multiple times in the place it's used. So don't worry about that.

Some future expansion could be storing the replays in memory via like a cache in this system, so if a player reloads a map it'll just fetch from cache again instead of doing the requests. Same for individual replay files.

Gocnak commented on mp/src/game/shared/momentum/run/mom_replay_factory.h:
```diff
@@ -17,8 +24,50 @@ class CMomReplayFactory
+        MomReplayAsyncHandle                handle;
+        MomLocalReplayCallback_t callback;
+        FSAsyncControl_t*                   arrControls;    // We keep these so we can abort the requests if needed
+        size_t                              controls_count;
+        size_t                              controls_pending; // How many files have we not heard back from yet?
```
Should these be atomics? `CInterlockedInt` ?

bltech replied:
Yeah they should be, when looking at it, I just noticed I unlocked the mutex too soon in FilesystemCallback so the callback could in theory be called twice, that should be solved by making them atomic.

Gocnak commented on mp/src/game/shared/momentum/run/mom_replay_factory.h:
```diff
@@ -17,8 +24,50 @@ class CMomReplayFactory
+    {
+        MomReplayAsyncHandle                handle;
+        MomLocalReplayCallback_t callback;
+        FSAsyncControl_t*                   arrControls;    // We keep these so we can abort the requests if needed
```
Why not also a CUtlVector? Or a CUtlLinkedList?

bltech replied:
When looking at filesystem.h for too long you really get caught up in raw C, I'll switch this to containers.

Gocnak commented on mp/src/game/shared/momentum/run/mom_replay_factory.h:
```diff
@@ -17,8 +24,50 @@ class CMomReplayFactory
+    struct MomReplayLoadRequestInternal
+    {
+        MomReplayAsyncHandle                handle;
+        MomLocalReplayCallback_t callback;
```
Spacing

Gocnak commented on mp/src/game/shared/momentum/mom_shareddefs.h:
```diff
@@ -185,6 +185,17 @@ const char* const g_szTimesStatusStrings[] = {
     "#MOM_API_UnauthFriendsList", // STATUS_UNAUTHORIZED_FRIENDS_LIST
 };
 
+enum LocalTimesStatus_t
+{
+    LOCALREPLAY_READY,               // Everything was loaded correctly and is ready to be read
+    LOCALREPLAY_INPROGRESS,          // Filesystem is still busy
+    LOCALREPLAY_CANCELLED,               // The request was manually cancelled
+    LOCALREPLAY_BADREQUEST,               // The request was ill-formed (non-existent PathID)
+    LOCALREPLAY_FS_ERROR    // Every file failed due to a filesystem error
+};
+
+DECLARE_POINTER_HANDLE(MomReplayAsyncHandle);
+
```
Should move into an ifdef CLIENT_DLL inside of the replay factory as this is only used by that factory, and therefore anything using it would be including it. Shareddefs is more like, project-globals typically.

Gocnak commented on mp/src/game/shared/momentum/mom_shareddefs.h:
```diff
@@ -185,6 +185,17 @@ const char* const g_szTimesStatusStrings[] = {
     "#MOM_API_UnauthFriendsList", // STATUS_UNAUTHORIZED_FRIENDS_LIST
 };
 
+enum LocalTimesStatus_t
```
Also, this enum should probably be called something more generic like `AsyncReplayStatus_t` and start with `ASYNC_REPLAYS_` .

Gocnak commented on mp/src/game/shared/momentum/run/mom_replay_factory.cpp:
```diff
@@ -98,4 +120,159 @@ CMomReplayBase* CMomReplayFactory::LoadReplayFile(const char* pFileName, bool bF
+
+    intReq->replayMutex.Lock();
+    vecToReturn.AddVectorToTail(intReq->replays);
+    intReq->replays.Purge();
```
Needed to be done inside this mutex lock? Doesn't AsyncCleanup handle this anyways?

Gocnak commented on mp/src/game/shared/momentum/run/mom_replay_factory.cpp:
```diff
@@ -98,4 +120,159 @@ CMomReplayBase* CMomReplayFactory::LoadReplayFile(const char* pFileName, bool bF
+        intReq->callback();
+    }
+}
+LocalTimesStatus_t CMomReplayFactory::AsyncGetReplays(MomReplayAsyncHandle handle,
+                                                             CUtlVector<CMomReplayBase*> &vecToReturn,
+                                                             bool bCancelRemaining)
```
I feel like this is doing a bit too much. Canceling the remaining upon trying to read this just doesn't make sense to me as a use case. You either want them all when they're ready, or you cancel the requests (as you shutdown map / the game).

I mean maybe you can make an overload to the Cancel that's like `CancelAndFillPartial` but I am still trying to understand when this would be the desired case.

Gocnak commented on mp/src/game/shared/momentum/run/mom_replay_factory.cpp:
```diff
@@ -98,4 +120,159 @@ CMomReplayBase* CMomReplayFactory::LoadReplayFile(const char* pFileName, bool bF
+    AsyncCleanup(handle);
+
+    return LocalTimesStatus_t::LOCALREPLAY_READY;
+}
```
Spacing, add a new line between every function definition.

Gocnak commented on mp/src/game/shared/momentum/run/mom_replay_factory.cpp:
```diff
@@ -98,4 +120,159 @@ CMomReplayBase* CMomReplayFactory::LoadReplayFile(const char* pFileName, bool bF
+    internalRequest->controls_count = count;
+    internalRequest->controls_pending = count;
+
+    FOR_EACH_VEC (filenames, i)
```
Since this one's a macro you won't want spacing between the macro name and the params.

Gocnak commented on mp/src/game/shared/momentum/run/mom_replay_factory.cpp:
```diff
@@ -98,4 +120,159 @@ CMomReplayBase* CMomReplayFactory::LoadReplayFile(const char* pFileName, bool bF
+{
+    g_ReplayFactory.FilesystemCallback(request, nBytesRead, err);
+}
+void CMomReplayFactory::FilesystemCallback(const FileAsyncRequest_t &request, int nBytesRead, FSAsyncStatus_t err)
```
`err` is kinda misleading as a variable name, prefer `status`:
```suggestion
void CMomReplayFactory::FilesystemCallback(const FileAsyncRequest_t &request, int nBytesRead, FSAsyncStatus_t status)
```

bltech replied:
It's basically copy-paste from the typedef from filesystem.h but I agree

Gocnak commented on mp/src/game/client/momentum/ui/leaderboards/ClientTimesDisplay.cpp:
```diff
@@ -201,6 +201,8 @@ void CClientTimesDisplay::FireGameEvent(IGameEvent *event)
     if (FStrEq(type, "replay_save") && event->GetBool("save"))
```
Hmm. Should it reload *every* replay if you only made one new one?

I believe the replay path is passed along here. Could be very possible to async load just that one and add it? Can use the same paths mostly but would save on some threading and compute.

Gocnak replied:
I say this because `hud_mapfinished` and the run uploader also load this same replay, but non-asynchronously. A future branch could be using this code you'd add here for them.

bltech replied:
That would make sense however I'm not really sure how this would tie in to your proposal around the top where the AsyncLoadMultiple method would only get the map name. If it was a wildcard that would work but the callback structure would need to be changed a bit.

Following this logic, OnConfirmDeleteReplay should also maybe not reload everything and only remove the entry after deleting the file. But on the other hand going through all CMomReplayBase* objects and somehow associating them to the deleted file might not be worth the hassle.

Gocnak replied:
Right -- it would be another function like `LoadReplayByFileAsync` and your current one would become `LoadReplaysForMapAsync`. Both of these should call into different callbacks but would ultimately do roughly the same thing of fetching the file via handle, handling that file (inserting and converting) then clearing the request.

I agree that OnConfirmDeleteReplay should only loop through currently loaded ones and delete it from the list and re-sort. It was also pretty naively implemented by just reloading everything.

Something to note is that those message functions can very easily be`MESSAGE_FUNC_PARAMS` that take the keyvalues in, which can hold pointers tied to a string key.

Gocnak commented on mp/src/game/client/momentum/ui/leaderboards/LeaderboardsTimes.h:
```diff
@@ -75,13 +79,15 @@ class CLeaderboardsTimes : public vgui::EditablePanel
     void SetPlaceColors(vgui::SectionedListPanel* panel, TimeType_t type) const;
-    void LoadLocalTimes(KeyValues *kv);
+    void LoadLocalTimes();
+    void LocalTimesCallback();
     void LoadOnlineTimes(TimeType_t type);
     void ConvertLocalTimes(KeyValues *pKv);
+    void LocalTimesVectorToLeaderboards();
     // Place vector times into leaderboards panel (sectionlist)
     void OnlineTimesVectorToLeaderboards(TimeType_t type);
 
-    bool GetPlayerTimes(KeyValues *outPlayerInfo, bool fullUpdate);
+    bool FillPlayerTimes(bool fullUpdate);
```
Since there is now 4 local times related methods, try seeing if you can move some to be private. `ConvertLocalTimes` followed by a public `LocalTimesVectorToLeaderboards` is a little weird. Can some be combined, especially if it's only used by one function?

Gocnak commented on mp/src/game/client/momentum/ui/leaderboards/LeaderboardsTimes.cpp:
```diff
@@ -264,38 +265,17 @@ int CLeaderboardsTimes::TryAddAvatar(const uint64 &steamid, CUtlMap<uint64, int>
-        SetVisible(false);
-
-        if (m_pCurrentLeaderboards == m_pLocalLeaderboards)
+        if(m_pCurrentLeaderboards == m_pLocalLeaderboards)
```
Spacing

Gocnak reviewed (changes requested):
Stepping in the right direction!

bltech commented:
I (hope) all requests are adressed. 2 Things to note: 
- I moved the initial replay request to LevelInit since it is supposed to get called once anyways, which has the bonus of the panel being ready when opened for the first time even with lots of replays (loading times should not increase according to my tests), which I could change if this is really not desired.
- Secondly, I switched to just 1 mutex that blocks basically the whole critical functions. While this is in theory not optimal, there were just too many edge cases to consider regarding deadlocks and Time-of-Check-to-Time-of-Use-Problems. I wasn't able to measure any performance impact of this change. This is especially true since it seems that the filesystem.h functions don't invoke callbacks simultaneously anyways and the only point of contention would be closing a level.

Gocnak commented on mp/src/game/client/momentum/ui/leaderboards/LeaderboardsTimes.cpp:
```diff
@@ -979,17 +954,36 @@ void CLeaderboardsTimes::ApplySchemeSettings(IScheme* pScheme)
+
+        bool bFound = false;
+
+        char filenameInVec[MAX_PATH];
+        FOR_EACH_VEC(m_vLocalTimes, i)
+        {
+
+            Q_snprintf(filenameInVec, MAX_PATH, "%s/%s-%s%s", RECORDING_PATH, m_vLocalTimes[i]->GetMapName(),
+                       m_vLocalTimes[i]->GetRunHash(), EXT_RECORDING_FILE);
+
+            if (FStrEq(filenameInVec, file))
+            {
+                m_vLocalTimes.Remove(i);
+                bFound = true;
+                break;
+            }
+        }
+        if (!bFound)
         {
-            g_pMOMRunCompare->LoadComparisons();
-            FillLeaderboards(false);
+            Warning("Error while deleting replay: File not found! This probably means that the replay file name was invalid.");
+            return;
         }
+        // MOM_TODO this is still synchoronous so it will lag
+        g_pMOMRunCompare->LoadComparisons();
+        FillLeaderboards(false);
```
I can do you one better:

`ConvertLocalTimes` should store the replay pointer inside of the item data via something like:
`kvLocalTimeFormatted->SetPtr("replayPtr", t);`

You can then feed this into the ConfirmDeleteReplay context menu item in `OnItemContextMenu` since it fetches this exact KV inside the variable `selectedRun`. You should still pass `runName` and `itemID` as those will handle removing it from the list of runs on the replay and deleting the file, but you should also add a nice forward:
`pMessage->SetPtr("replayPtr", selectedRun->GetPtr("replayPtr"));`

You can then change the type of the `ContextDeleteReplay` message to be:
`MESSAGE_FUNC_PARAMS(OnContextDeleteReplay, "ContextDeleteReplay", pData);`

And then, finally, in this function, do:
`m_vLocalTimes.FindAndRemove(pData->GetPtr("replayPtr"));`

You'll also need to convert the function to take the `KeyValues *pData` from the macro, and then get all of the data manually (e.g. `const char *file = pData->GetString("runName");` ) but it will be much cleaner in the end.

Gocnak replied:
The reason to do this is to tie in with my previous comment about the `RemoveAll`, because I don't think that's a good idea unless the map is being shut down.

bltech replied:
I'll implement this however one small thing, `runName` should really go because it gets constructed via the mapname-hast.mrf logic in ConvertLocalTimes

bltech replied:
Also I've run into another issue, mom_replay_play expects the path without the replays/ part. 
I've settled on setting runName to `replayPtr->GetFilePath()[Q_strlen(RECORDING_PATH)]` since we are on the local panel and there is no way for a file not to be in that folder, unless you have a better idea.

Gocnak replied:
Yeah that's fine.

Gocnak commented on mp/src/game/shared/momentum/run/mom_replay_factory.h:
```diff
@@ -2,9 +2,31 @@
 class CMomReplayBase;
 
+#ifdef CLIENT_DLL
+#include <tier1/utlpair.h>
```
Needed include?

Gocnak commented on mp/src/game/shared/momentum/run/mom_replay_factory.h:
```diff
@@ -17,8 +39,55 @@ class CMomReplayFactory
+        CUtlVector<CMomReplayBase*>         replays;
+    };
+
+    CUtlVector<MomReplayLoadRequestInternal> requests;
```
```suggestion
    CUtlVector<MomReplayLoadRequestInternal> m_vecRequests;
```

Gocnak commented on mp/src/game/shared/momentum/run/mom_replay_factory.h:
```diff
@@ -17,8 +39,55 @@ class CMomReplayFactory
+    };
+
+    CUtlVector<MomReplayLoadRequestInternal> requests;
+    CThreadMutex requestMutex; // While the documentation is sparse, this is a recursive mutex
```
```suggestion
    CThreadMutex m_requestMutex; // While the documentation is sparse, this is a recursive mutex
```

Gocnak commented on mp/src/game/shared/momentum/run/mom_replay_factory.cpp:
```diff
@@ -98,4 +108,201 @@ CMomReplayBase* CMomReplayFactory::LoadReplayFile(const char* pFileName, bool bF
+
+    do
+    {
+        // This is not 100% foolproof validation but it should be enough unless you purposefully break it
+        if (Q_strlen(pFoundFile) == Q_strlen(mapName) + 1 /* - */ + REPLAY_HASH_LENGTH + Q_strlen(EXT_RECORDING_FILE))
```
Okay so after a bit of reading the original code I understand why it breaks. The `ConvertLocalTimes` guy automatically assumes that the filename of the read replay *had* to be `mapname-hash.mrf`, but isn't always the case with the files you were messing around with.

Unfortunately as I found out, downloaded online replays have the format `mapname-runID.mrf` with file name lengths substantially smaller than the SHA1 hash ones.

So here's what I propose. Inside of `CMomReplayBase` we add a new variable for the file path it was read from. So far our use case for replays is strictly file-based, even ones from online leaderboards are locally downloaded as to save our servers the bandwidth. This can be a `char m_szFilePath[MAX_PATH];` right below the hash of the replay in the private section. Then, you would want to expose a public virtual getter and setter for this file path.

Then, for everything that loads a replay from file, you can do something like `replay.SetFilePath(filepath);`. For this case, it'd be in `FileSystemCallback`, right before it gets added to the vector, and you can get the file name from the `FileAsyncRequest_t` object.

This can then let us get the exact file name of the replay from the pointer of it, and thus it doesn't matter what it's called. The `ConvertLocalTimes` guy would instead use the public getter from the pointer instead of reconstructing it and making assumptions. You can also use it directly in my previous suggestion with regards to the `ContextDeleteReplay` stuff. You can also then safely remove this check as it won't matter what it's called.

Gocnak commented:
It's a large card so no worries, code review is part of the process. :)

bltech commented on mp/src/game/client/momentum/ui/leaderboards/LeaderboardsTimes.cpp:
```diff
@@ -311,10 +293,31 @@ void CLeaderboardsTimes::FillLeaderboards(bool bFullUpdate)
+    if (kvLocalTimes->IsEmpty())
+        return;
+
+    m_pLocalLeaderboards->RemoveAll();
```
Since the keyvalues are gonna hold the replay pointer anyways, maybe m_vLocalTimes can be replces with a keyvalue completely?

Gocnak replied:
You could make m_vLocalTimes be a vector of keyvalues yeah, but note that ConvertTimes needs to be called on any replay being loaded.

Gocnak replied:
I think with my suggestion at https://github.com/momentum-mod/game/pull/1061#discussion_r515472830 you'll want to keep the vector of replays itself.

bltech commented:
I'm almost done now, the only issue I can find is that if you have the leaderboards open while submitting a run, the place colors can get messed up but after reopening they are fine. Is there some smart way to do whatever the panel does when it reopens? I tried InvalidateLayout on the Leaderboards, on the SectionedList but nothing worked.

Gocnak commented:
So the problem is when the individual run (run_save event) is added, place colors aren't updated? Could you call SetPlaceColors manually for it? I thought it would be going down the code path for it already though.

bltech commented:
SetPlaceColors is getting called but it seems like it's not applying right away. If I keep it open long enough, eventually the new color gets added but only when reopening, the other colors disappear. I also tried to set every following item to panel but that still only helps once the 15 sec update timer of clienttimesdisplay hits

bltech commented:
I think it might be because the sectionlist doesn't resort right away?

Gocnak commented:
I think if you push it for now I can step through it; I'm trying to see locally on develop's code but you've deviated from it.

My hunch is, opening the panel calls `Reset` which in the old code went and added / removed all the replays again, but in your code path, it could be missing something key with adding the individual replay, but I have to see it before I can determine.

Gocnak commented on mp/src/game/shared/momentum/run/mom_replay_factory.cpp:
```diff
@@ -70,6 +64,22 @@ CMomReplayBase* CMomReplayFactory::LoadReplayFile(const char* pFileName, bool bF
         return nullptr;
     }
 
+    return LoadReplayFromBuffer(reader, bFullLoad);
```
Need to set the file path here:

```suggestion
    const auto pReplay = LoadReplayFromBuffer(reader, bFullLoad);
    if (pReplay)
        pReplay->SetFilePath(pFileName);
    return pReplay;
```

bltech commented:
There you go

Gocnak commented:
It's only the adding that's causing this issue, not the removing right?

bltech commented:
Yeah deleting works fine

Gocnak commented:
So I would suggest to try the `InvalidateItem` method call for the item being added.

Opening/closing the leaderboards still loops through every item after deleting and re-adding them, so that's why it works there. There's some choice work to do in SectionedListPanel but I think it's heavily out of scope and this bug shouldn't bother you currently or delay this PR further.

Gocnak reviewed (changes requested):
Yep this looks great, just one minor thing then it's merge time

bltech commented:
Unfortunately InvalidateItems does nothing either, I suspect it's because it calls InvalidateLayout with layoutNot= false which doesn't repaint but that's really for another time, I agree.
